Add specialized implementations of fixed stream message to reduce all…

anuraaga opened the pull request:
…ocations for common cases.

It's very common to allocate stream messages with 0-2 objects, so it's nice to have specializations to avoid the extra garbage of an object array. 

Unfortunately, the duplicate implementations introduce a lot of complexity. Let me know if you think it's worth it. `FixedHttpRequest` turned out much worse than I had originally expected.

The total allocation of each is
- Empty: 16 bytes (3 fields in base class)
- One element: 16 bytes (3 fields in base, one in sub)
- Two elements: 24 bytes (3 fields in base, 3 in sub)
- Three elements: 24 + 36 (3 fields in base, 3 in sub, 24 bytes array overhead, 12 bytes in array)

Also unrelated, removed `invokedOnSubscribe` as it's unnecessary for fixed streams which have no writes.

Benchmarks. Not totally apple-to-apple due to the `switch` overhead in `After` from what I can tell - I think the overhead is shown by `num=3`, where the implementation hasn't really changed, so presumably `After` should be treated a little higher than they look. But still pretty minor, let me know what you think.

```
After

Benchmark                            (flowControl)  (num)          (streamType)   Mode  Cnt         Score        Error  Units
StreamMessageBenchmark.jmhEventLoop          false      0  FIXED_STREAM_MESSAGE  thrpt   20  11644168.631 ± 152677.634  ops/s
StreamMessageBenchmark.jmhEventLoop          false      1  FIXED_STREAM_MESSAGE  thrpt   20   8573895.575 ±  34659.073  ops/s
StreamMessageBenchmark.jmhEventLoop          false      2  FIXED_STREAM_MESSAGE  thrpt   20   6602223.767 ±  24179.698  ops/s
StreamMessageBenchmark.jmhEventLoop          false      3  FIXED_STREAM_MESSAGE  thrpt   20   5326205.531 ±  17208.362  ops/s

Before

Benchmark                            (flowControl)  (num)          (streamType)   Mode  Cnt         Score       Error  Units
StreamMessageBenchmark.jmhEventLoop          false      0  FIXED_STREAM_MESSAGE  thrpt   20  11338855.960 ± 56332.436  ops/s
StreamMessageBenchmark.jmhEventLoop          false      1  FIXED_STREAM_MESSAGE  thrpt   20   8165172.831 ± 22656.208  ops/s
StreamMessageBenchmark.jmhEventLoop          false      2  FIXED_STREAM_MESSAGE  thrpt   20   6681994.146 ± 32692.861  ops/s
StreamMessageBenchmark.jmhEventLoop          false      3  FIXED_STREAM_MESSAGE  thrpt   20   5441368.221 ± 20874.088  ops/s

```

trustin commented on core/src/main/java/com/linecorp/armeria/common/FixedHttpRequest.java:
```diff
@@ -19,64 +19,191 @@
     /**
-     * Creates a new {@link FixedHttpRequest} that will publish the given {@code objs}. {@code objs} is not
+     * Creates a new {@link HttpRequest} that will publish the given {@code objs}. {@code objs} is not
      * copied so must not be mutated after this method call (it is generally meant to be used with a varargs
```
I guess this may be premature optimization. Given this is not true for the case there `objs.length` <= 2, could we just make `RefularFixedHttpRequest` make a copy of `objs`?

trustin commented on core/src/main/java/com/linecorp/armeria/common/stream/FixedStreamMessage.java:
```diff
@@ -33,277 +35,467 @@
     /**
-     * Creates a new {@link FixedStreamMessage} that will publish the given {@code objs}. {@code objs} is not
+     * Creates a new {@link StreamMessage} that will publish the given {@code objs}. {@code objs} is not
      * copied so must not be mutated after this method call (it is generally meant to be used with a varargs
```
Ditto

anuraaga commented:
Thanks a lot for the suggestion! Indeed, there was significant duplication between those factories and the one in `HttpRequest`, and there's no good point for it. I made `FixedHttp*` package-private, and also fixed my misunderstanding that it's possible to have a `HttpRequest` that publishes headers (and made an explicit check to prevent this).

I also went ahead and deprecated `AggregatedHttpMessage.to*` in the interest of making `HttpRequest` and `HttpResponse` the only entrypoints for creating them (will make the API clearer after #850 ). In the process, I moved some error-checking logic from there to `HttpRequest/Response`, presumably it was supposed to be there either way.

Also, what do you think about moving `FixedStreamMessage` to `armeria.internal.stream`? I would only do it if we're also ok with moving `Default/EventLoopStreamMessage` there to later.

trustin commented:
> Also, what do you think about moving `FixedStreamMessage` to `armeria.internal.stream`? I would only do it if we're also ok with moving `Default/EventLoopStreamMessage` to there later.

What's the benefit of moving them there? I usually prefer keeping classes package-local if possible since the classes in the internal package are visible to users anyway.

trustin commented on core/src/main/java/com/linecorp/armeria/common/AggregatedHttpMessage.java:
```diff
@@ -429,74 +429,20 @@ default HttpStatus status() {
      *
-     * @return the converted {@link HttpRequest} whose stream is complete
-     * @throws IllegalStateException if this message is not a request
+     * @deprecated use {@link HttpRequest#of(AggregatedHttpMessage)}.
```
use -> Use

trustin commented on core/src/main/java/com/linecorp/armeria/common/AggregatedHttpMessage.java:
```diff
@@ -429,74 +429,20 @@ default HttpStatus status() {
      *
-     * @return the converted {@link HttpResponse} whose stream is complete
-     * @throws IllegalStateException if this message is not a response.
+     * @deprecated use {@link HttpResponse#of(AggregatedHttpMessage)}.
```
use -> Use

trustin commented on core/src/main/java/com/linecorp/armeria/common/stream/FixedStreamMessage.java:
```diff
@@ -33,277 +36,465 @@
+        private boolean inOnNext;
+
+        protected RegularFixedStreamMessage(T[] objs) {
+            this.objs = Arrays.copyOf(objs, objs.length);
```
`objs.clone()`?

trustin commented on core/src/main/java/com/linecorp/armeria/common/stream/FixedStreamMessage.java:
```diff
@@ -33,277 +36,465 @@
+            case 2:
+                return of(objs[0], objs[1]);
+            default:
+                return new RegularFixedStreamMessage<>(objs);
```
Check all elements are non-null?

trustin commented on core/src/main/java/com/linecorp/armeria/common/stream/FixedStreamMessage.java:
```diff
@@ -33,277 +36,465 @@
  * so this class should generally be used when the objects are known.
  */
-public class FixedStreamMessage<T> extends AbstractStreamMessage<T> {
+public final class FixedStreamMessage {
```
How about:

- moving all the factory methods to `StreamMessage`
- removing `FixedStreamMessage` and moving all its nested classes top-level
- renaming `AbstractFixedStreamMessage` to `FixedStreamMessage`?

anuraaga replied:
Thanks - reorganized. Unfortunately I can't make the `FixedStreamMessage` implementations package-private since we inherit from outside the package. It's why I was thinking of moving them to `internal` but I guess this way is ok for now.

minwoox commented on core/src/main/java/com/linecorp/armeria/common/stream/FixedStreamMessage.java:
```diff
@@ -16,24 +16,21 @@
- * A {@link StreamMessage} used when all the objects that will be published are known at construction time.
- * Reduced synchronization and allocation allow for much higher performance than {@link DefaultStreamMessage},
- * so this class should generally be used when the objects are known.
+ * A {@link AbstractStreamMessage} which only publishes a fixed number of objects known at construction time.
```
nit: A {@link Fixed~~ } or An {@link Abstract~~~}

anuraaga replied:
Thanks

trustin commented:
Thanks, @anuraaga !
